sw/device/lib/testing/mock_mmio.h

```diff
@@ -7,14 +7,112 @@
 
 #include <stdint.h>
 
+#include <initializer_list>
 #include <memory>
+#include <random>
 #include <vector>
 
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
 #include "sw/device/lib/base/mmio.h"
 
 namespace mock_mmio {
+/**
+ * Represents a single bit field in an integer, useable with EXPECT_* macros
+ * defined in this file.
+ *
+ * An integer can be expressed as a list of BitField values, and can be more
+ * convenient to use than 0b or 0x literals in some cases. For example, the
+ * integer 0b0000'0000'1100'0101 could be expressed as
+ *   {{0x0, 1}, {0x2, 1}, {0x4, 12}}
+ * This form makes it clearer to the reader that 0x0, 0x2, and 04 are indices
+ * to bitfields, which are set to particular values.
+ *
+ * In practice, this might use generated register constants, and look like
+ *   {{FIELD_FOO_OFFSET, 1}, {FIELD_BAR_OFFSET, 1}, {FIELD_BAZ_OFFSET, 12}}
+ *
+ * This type does not specify the lengths of bitfields; MaskedBitField should be
+ * used for that, instead.
+ */
+struct BitField {
+  uintptr_t offset;
+  uintptr_t value;
+};
+
+/**
+ * Represents a single bit field in an integer, similar to BitField. It can be
+ * used in most places that need a BitField, as well as in |EXPECT_MASK| macros.
+ *
+ * Like with BitFields, we can express the integer 0b0000'0000'1100'0101 as a
+ * list of BitFieldMasks:
+ *   {{0x0, 0x1, 1}, {0x1, 0x1, 0}, {0x2, 0x1, 1}, {0x3, 0x1, 0}, {0x4, 0xff,
+ * 12}} In addition to showing how the integer is broken up, it also expresses
+ * the lengths of fields, so it is clear that 0x0 and 0x2 are one-bit fields.
+ * This also allows us to formally express that the fields 0x1 and 0x3 are
+ * *unset*.
+ *
+ * In practice, this might use generated register constants, and look like
+ *   {{FIELD_FOO_OFFSET, FIELD_FOO_MASK, 1}, {FIELD_BAR_OFFSET, FIELD_BAR_MASK,
+ * 0}, ...}
+ */
+struct MaskedBitField {
+  uintptr_t offset;
+  uintptr_t mask;
+  uintptr_t value;
+
+  BitField ToBitField() {
+    return {offset, value};
+  }
+};
+
+namespace internal {
+/**
+ * Implicit conversion guard around |char *|.
+ */
+struct WrappedBytes {
+  const char *bytes;
+};
+
+/**
+ * Internal overloaded function for converting various types into little-endian
+ * unsigned integers of various widths.
+ */
+template <typename Int>
+Int ToLeInt(Int val) {
```

Comment:
I worry these conversions are fraught, especially with regards to `WrappedBytes` carrying a little-endian number always.  Why not types that encode their endianness? `LittleEndian<uint32_t>`, for example?

Reply:
I wrote a bunch of documentation, like we discussed. Hopefully that sheds some light on why this is written the way it is.